File: divedave-ios/divedave Shared/Components/Scene/DiveScorer.swift
import CoreGraphics
import Foundation

struct DiveOutcome {
    enum Result { case success, failure }
    let result: Result
    let scores: [Int]
    let emotionFrame: Int
}

enum BoostTiming {
    case perfect, good, ok, miss
}

enum DiveScorer {
    static func classifyBoostTiming(quicknessMs: Double) -> BoostTiming {
        if quicknessMs < Game.boostPerfectMs { return .perfect }
        if quicknessMs < Game.boostGoodMs { return .good }
        if quicknessMs < Game.boostOkMs { return .ok }
        return .miss
    }

    static func score(goalRotations: Double, rotations: Double, angle: Double, tuckCount: Int) -> DiveOutcome {
        guard abs(rotations - goalRotations) < 0.25 else {
            return DiveOutcome(result: .failure, scores: [0, 0, 0], emotionFrame: 0)
        }

        let baseFrame = chooseEmotionFrame(angle: angle)

        var scores: [Int] = []
        for _ in 0..<3 {
            let s: Int
            switch baseFrame {
            case 4: s = Int(10 - Double(Int.random(in: 0...1)) / 2.0 - Double(tuckCount - 1))
            case 3: s = Int(10 - Double(Int.random(in: 3...6)) / 2.0 - Double(tuckCount - 1))
            case 2: s = Int(10 - Double(Int.random(in: 7...10)) / 2.0 - Double(tuckCount - 1))
            case 1: s = Int(10 - Double(Int.random(in: 10...15)) / 2.0 - Double(tuckCount - 1))
            case 0: s = Int(10 - Double(Int.random(in: 14...18)) / 2.0 - Double(tuckCount - 1))
            default: s = 0
            }
            scores.append(s)
        }

        let adjustedFrame = tuckCount > 1 ? baseFrame - 1 : baseFrame
        return DiveOutcome(result: .success, scores: scores, emotionFrame: adjustedFrame)
    }

    static func chooseEmotionFrame(angle: Double) -> Int {
        let absAngle = abs(angle)

        let boundary10 = 10.0
        let boundary25 = 25.0
        let boundary45 = 45.0
        let boundary70 = 70.0
        let boundary110 = 110.0
        let boundary135 = 135.0
        let boundary155 = 155.0
        let boundary170 = 170.0

        if absAngle < boundary10 || absAngle > boundary170 {
            return 4
        } else if (absAngle >= boundary10 && absAngle < boundary25) || (absAngle <= boundary170 && absAngle > boundary155) {
            return 3
        } else if (absAngle >= boundary25 && absAngle < boundary45) || (absAngle <= boundary155 && absAngle > boundary135) {
            return 2
        } else if (absAngle >= boundary45 && absAngle < boundary70) || (absAngle <= boundary135 && absAngle > boundary110) {
            return 1
        } else if absAngle >= boundary70 && absAngle < boundary110 {
            return 0
        }
        return 2
    }

    static func heightInMeters(springboardY: CGFloat, waterY: CGFloat) -> Double {
        let heightDifference = springboardY - waterY
        let inMeters = heightDifference / 200.0
        return round(Double(inMeters) * 10) / 10
    }

    // Dive height (metres) → number of half-flips the goal may ask for. A tuning
    // heuristic, not real physics; computed in reference-screen space so the goal
    // is device-independent. Mirrors divedave-web goalHalfFlips (parity-tested).
    static func goalHalfFlips(heightMeters: Double) -> Int {
        let referenceScale = Double(Game.referenceScreenHeight) / Double(Game.defaultHeight)
        let diveHeight = heightMeters * 200.0 * referenceScale
        let waterLevel = 256.0 * referenceScale / 2.0 + 1.0
        let distance = max(0.0, diveHeight - waterLevel)
        let time = sqrt(distance) / 10.0
        let totalRotation = time * Double(Game.maxSpinVelocity) * 0.70
        return Int((totalRotation / (2.0 * .pi)) * 2.0)
    }
}

File: divedave-web/src/components/scene/DiveScorer.js
import {
  BOOST_GOOD_MS,
  BOOST_OK_MS,
  BOOST_PERFECT_MS,
  MAX_SPIN_VELOCITY,
} from "../../util/Constants.js";
import { getRandomIntInclusive } from "../../util/Utilities.js";

export const DiveResult = Object.freeze({
  Success: "success",
  Failure: "failure",
});

export const BoostTiming = Object.freeze({
  Perfect: "perfect",
  Good: "good",
  Ok: "ok",
  Miss: "miss",
});

export function classifyBoostTiming(quicknessMs) {
  if (quicknessMs < BOOST_PERFECT_MS) return BoostTiming.Perfect;
  if (quicknessMs < BOOST_GOOD_MS) return BoostTiming.Good;
  if (quicknessMs < BOOST_OK_MS) return BoostTiming.Ok;
  return BoostTiming.Miss;
}

export function scoreDive({ goalRotations, rotations, angle, tuckCount }) {
  if (Math.abs(rotations - goalRotations) >= 0.25) {
    return { result: DiveResult.Failure, scores: [0, 0, 0], emotionFrame: 0 };
  }

  const baseFrame = chooseEmotionFrame(angle);
  const scores = [0, 1, 2].map(() => scoreForFrame(baseFrame, tuckCount));

  const emotionFrame = tuckCount > 1 ? baseFrame - 1 : baseFrame;
  return { result: DiveResult.Success, scores, emotionFrame };
}

function scoreForFrame(frame, tuckCount) {
  switch (frame) {
    case 4:
      return Math.trunc(10 - getRandomIntInclusive(0, 1) / 2.0 - (tuckCount - 1));
    case 3:
      return Math.trunc(10 - getRandomIntInclusive(3, 6) / 2.0 - (tuckCount - 1));
    case 2:
      return Math.trunc(10 - getRandomIntInclusive(7, 10) / 2.0 - (tuckCount - 1));
    case 1:
      return Math.trunc(10 - getRandomIntInclusive(10, 15) / 2.0 - (tuckCount - 1));
    case 0:
      return Math.trunc(10 - getRandomIntInclusive(14, 18) / 2.0 - (tuckCount - 1));
    default:
      return 0;
  }
}

export function chooseEmotionFrame(angle) {
  const a = Math.abs(angle);
  if (a < 10 || a > 170) return 4;
  if ((a >= 10 && a < 25) || (a <= 170 && a > 155)) return 3;
  if ((a >= 25 && a < 45) || (a <= 155 && a > 135)) return 2;
  if ((a >= 45 && a < 70) || (a <= 135 && a > 110)) return 1;
  if (a >= 70 && a < 110) return 0;
  return 2;
}

export function heightInMeters(springboardY, waterY) {
  const heightDifference = waterY - springboardY;
  const inMeters = heightDifference / 200.0;
  return Math.round(inMeters * 10) / 10;
}

// Dive height (metres) → number of half-flips the goal may ask for. A tuning
// heuristic, not real physics; computed in iOS's reference-screen space (852/3000)
// so the curve is identical to native. Mirrors DiveScorer.swift goalHalfFlips
// (parity-tested); the 852/3000 must track iOS Game.referenceScreenHeight.
export function goalHalfFlips(heightMeters) {
  const referenceScale = 852 / 3000;
  const diveHeight = heightMeters * 200 * referenceScale;
  const waterLevel = (256 * referenceScale) / 2 + 1;
  const distance = Math.max(0, diveHeight - waterLevel);
  const time = Math.sqrt(distance) / 10;
  const totalRotation = (time * MAX_SPIN_VELOCITY * 0.7 * Math.PI) / 180;
  return Math.floor((totalRotation / (2 * Math.PI)) * 2);
}
